torch/csrc/dynamo/guards.cpp

```
@@ -674,6 +674,621 @@ static struct PyModuleDef _module = {
+    return check_verbose_nopybind(value.ptr());
+  }
+
+  virtual GuardDebugInfo check_verbose_nopybind(
+      PyObject* value) { // borrowed ref
+    bool result = check_nopybind(value);
+    if (!result) {
+      return GuardDebugInfo(result, _verbose_code_parts, 0);
+    }
+    return GuardDebugInfo(true, 0);
+  }
+
+  py::list verbose_code_parts() {
+    return _verbose_code_parts;
+  }
+
+  // This is on the hot path and avoids any refcounting code from pybind. This
+  // is not exposed to Python and can only be called from C++.
+  virtual bool check_nopybind(PyObject* value) = 0;
+  virtual ~LeafGuard() = default;
+
+ private:
+  // This is set while constructing the leaf guard. This is used for identifying
+  // the cause of recompilation.
+  py::list _verbose_code_parts;
+};
+
+/**
+ * Represents a leaf guard that accepts the python guard check function. We
+ * would like to have most of the guards in C++ (to avoid a Python function
+ * call).  But, it will take some time to reach that goal. Also, there might be
+ * cases where its too tedious to write an equivalent C++ guard.
+ *
+ * LAMBDA_GUARD allows us to gradually move to C++. We can start from all
+ * guards of type PythonLambaGuard and incrementally move expensive guards to
+ * C++.
+ */
+class LAMBDA_GUARD : public LeafGuard {
+ public:
+  LAMBDA_GUARD(py::object guard_check_fn, py::object verbose_code_parts)
+      : LeafGuard(verbose_code_parts) {
+    if (py::isinstance<py::function>(guard_check_fn)) {
+      _guard_check_fn = py::cast<py::function>(guard_check_fn);
+    } else {
+      throw py::type_error("LAMBDA_GUARD expects (callable, str)");
+    }
+  }
+
+  // Runs the lambda function with the current f_locals value.
+  bool check_nopybind(PyObject* value) override { // borrowed ref
+    PyObject* x = PyObject_CallOneArg(_guard_check_fn.ptr(), value); // new ref
+    if (x == nullptr) {
+      // An exception is caught in the lambda function.
```

If this is supposed to be unreachable should we assert rather than silently returning?

If we want to return you need to clear the Python error.

---

Clearing the expection and returning False. I don't want to assert because otherwise then we might not DECREF some other pointer held higher up in the GuardManager tree.